iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ObjectTypeUtils.java

```diff
@@ -40,6 +58,117 @@ public class ObjectTypeUtils {
+        relativePath, offset, actualReadSize, mayNotInCurrentNode);
+  }
+
+  public static ByteBuffer readObjectContent(
+      String relativePath, long offset, int readSize, boolean mayNotInCurrentNode) {
+    Optional<File> objectFile = TIER_MANAGER.getAbsoluteObjectFilePath(relativePath, false);
+    if (objectFile.isPresent()) {
+      return readObjectContentFromLocalFile(objectFile.get(), offset, readSize);
+    }
+    if (mayNotInCurrentNode) {
+      return readObjectContentFromRemoteFile(relativePath, offset, readSize);
+    }
+    throw new ObjectFileNotExist(relativePath);
+  }
+
+  private static ByteBuffer readObjectContentFromLocalFile(File file, long offset, long readSize) {
+    byte[] bytes = new byte[(int) readSize];
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    try (FileChannel fileChannel = FileChannel.open(file.toPath(), StandardOpenOption.READ)) {
+      fileChannel.read(buffer, offset);
+    } catch (IOException e) {
+      throw new IoTDBRuntimeException(e, TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode());
+    }
+    buffer.flip();
+    return buffer;
+  }
+
+  private static ByteBuffer readObjectContentFromRemoteFile(
+      final String relativePath, final long offset, final int readSize) {
+    ByteBuffer buffer = ByteBuffer.allocate(readSize);
+    TConsensusGroupId consensusGroupId =
+        new TConsensusGroupId(
+            TConsensusGroupType.DataRegion,
+            Integer.parseInt(Paths.get(relativePath).getName(0).toString()));
+    List<TRegionReplicaSet> regionReplicaSetList =
+        ClusterPartitionFetcher.getInstance()
+            .getRegionReplicaSet(Collections.singletonList(consensusGroupId));
+    TRegionReplicaSet regionReplicaSet = regionReplicaSetList.iterator().next();
+    final int batchSize = 1024 * 1024;
+    final TReadObjectReq req = new TReadObjectReq();
+    req.setRelativePath(relativePath);
+    for (int i = 0; i < regionReplicaSet.getDataNodeLocations().size(); i++) {
+      TDataNodeLocation dataNodeLocation = regionReplicaSet.getDataNodeLocations().get(i);
+      int toReadSizeInCurrentDataNode = readSize;
+      try (SyncDataNodeInternalServiceClient client =
+          Coordinator.getInstance()
+              .getInternalServiceClientManager()
+              .borrowClient(dataNodeLocation.getInternalEndPoint())) {
+        while (toReadSizeInCurrentDataNode > 0) {
+          req.setOffset(offset + buffer.position());
+          req.setSize(Math.min(toReadSizeInCurrentDataNode, batchSize));
+          toReadSizeInCurrentDataNode -= req.getSize();
+          ByteBuffer partial = client.readObject(req);
+          buffer.put(partial);
+        }
+      } catch (Exception e) {
+        logger.error("Failed to read object from datanode: {}", dataNodeLocation, e);
+        if (i == regionReplicaSet.getDataNodeLocations().size() - 1) {
+          throw new IoTDBRuntimeException(e, TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode());
```

Comment:
We've already record the real thread exception in warn log, here we only need a new error code and a error msg to notify users that this problem.